Music Spectrogram diffusion pipeline

kashif opened the pull request:
For issue #320  and #544

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_1044). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented:
@patil-suraj @anton-l maybe you could already take a look here if you find some time :-)

anton-l commented:
There's no `pipeline.__call__` yet, so I'll wait for @kashif's ping when it's ready :smile:

kashif commented:
yup getting there...

patrickvonplaten commented:
This looks like a great start! @kashif could you add a code snippet showing how to run the model for inference? Similar to https://github.com/huggingface/diffusers/pull/658#issue-1388250440 maybe? :-) 

Once we can reproduce some results locally, I think we'll have a much easier time getting this PR merged :-)

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,75 @@
+        if not return_dict:
+            return (mel,)
+
+        return MelPipelineOutput(mels=mel)
```
Shouldn't we maybe put the mel to sound model also in the pipeline?

kashif replied:
Yup that is the idea I wanted to get this running first and then tackle the Mel model which is currently available as a `KerasLayer` wrapper over a TF1 model... so will need to figure out how to port it over... The initial Keras -> onnx -> pytorch object route gave a slicing error...

patrickvonplaten commented on src/diffusers/models/t5_attention.py:
```diff
@@ -0,0 +1,421 @@
+        super().__init__()
+
+        self.token_encoder = TokenEncoder(t5config=t5config)
+        self.continuous_encoder = ContinuousEncoder(t5config=t5config)
```
I would actually seperate this into 3 model classes:

- 1) The `TokenEncoder` 
- 2) The `ContinuousEncoder` 
- 3) The `Decoder`

From my understanding only the `Decoder` is a diffusion model so I would say we move the classes `TokenEncoder` and `ContinuousEncoder` into `pipelien_spectrogram_diffusion.py` and then we code up the `Decoder` (maybe we can call it `T5Decoder`) in this file. 

Note that we cannot have a dependency on `transformers` in `src/diffusers/models` so I think we should code up the `Decoder` class from scratch here (copy paste `T5Block` from transformers. 

The `TokenEncoder` and `ContinuousEncoder` can heavily rely on `transformers` as it's totally fine to have a transformers dependency in the pipeline class.

patrickvonplaten replied:
Also I'd propose to not make use of `T5Config` for the `Decoder` class but instead just write out all the kwargs (just as is done here: https://github.com/huggingface/diffusers/blob/c3dcb6749b744bbf3ebfd1329de6a50009b849ee/src/diffusers/models/attention.py#L87 ) so that we can correctly inherit from `ModelMixin` and `ConfigMixin` (otherwise as you mentioned you get a weird `self.config.config` problem. 

For `TokenEncoder` and `ContinuousEncoder` I'd inherit from `transformer`'s `PreTrainedModel and then you can pass the `config` just as done here.

kashif replied:
got it!

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class SpectrogramDiffusionPipeline(DiffusionPipeline):
+    def __init__(self, cont_context_trans: ContinuousContextTransformer, scheduler: DDPMScheduler) -> None:
```
```suggestion
    def __init__(self, notes_encoder: SpectrogramNotesEnocder, continuous_encoder: SpectrogramContEncoder, decoder: T5FilmDecoder, scheduler: DDPMScheduler) -> None:
```

patrickvonplaten replied:
If possible could we split the three different models as shown in the diagram here: 
![68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6d757369632d73796e7468657369732d776974682d7370656374726f6772616d2d646966667573696f6e2f6172636869746563747572652e706e67](https://user-images.githubusercontent.com/23423619/200548760-d216cc33-41f4-435e-b8ff-0cc2ccfe1d7d.png)

into different classes? I think this makes the code a bit more agile and is easier to understand as well?

patrickvonplaten replied:
Also I think we should pass the mel-to-sound model here as well so that people can directly generate music with the pipeline. IIRC the open-source implementation of the mel-to-sound model was not super easy (only coded up in TF) or is there a nice PT version that matches 1-to-1? Let me know if you need help :-)

patrickvonplaten replied:
this pipeline class still needs to changed. Note that every independent components of the diagram should be it's own model as stated here: https://github.com/huggingface/diffusers/pull/1044/files#r1016473392

patrickvonplaten reviewed (commented):
Great progress overall! I think we're quite close to have finalized the API design :-) Let me know if something is not clear!

patrickvonplaten commented on scripts/convert_music_spectrogram_to_diffusers.py:
```diff
@@ -0,0 +1,71 @@
+    ]
+
+    gin_file = os.path.join(args.checkpoint_path, "..", "config.gin")
+    gin_config = inference.parse_training_gin_file(gin_file, gin_overrides)
```
Nice!

patrickvonplaten commented on scripts/convert_music_spectrogram_to_diffusers.py:
```diff
@@ -0,0 +1,71 @@
+
+
+def main(args):
+    t5_checkpoint = checkpoints.load_t5x_checkpoint(args.checkpoint_path)
```
I guess this checkpoint needs to be converted to PyTorch and loaded into the model somewhere no? :-) Looks very nice already though!

patrickvonplaten commented on src/diffusers/models/film.py:
```diff
@@ -0,0 +1,27 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
```
Could we maybe move this whole class into `t5_attention.py` ?

patrickvonplaten commented on scripts/convert_music_spectrogram_to_diffusers.py:
```diff
@@ -28,7 +28,9 @@ def load_token_encoder(weights, model):
         attention_weights = ly_weight["attention"]
-        lyr.layer[0].SelfAttention.q.weight = nn.Parameter(torch.FloatTensor(attention_weights["query"]["kernel"].T))
+        lyr.layer[0].SelfAttention.q.weight = nn.Parameter(
+            torch.FloatTensor(attention_weights["query"]["kernel"].T * depth_scaling)
```
Let's please not do this - weights should not be scaled just so that we don't show the scaling in the code. This will be very confusing for future weights that want to use this code (e.g. all future weights will also have to be scaled)

kashif replied:
yes I will revert it! this is not the issue

patrickvonplaten commented:
@kashif, let me know as soon as you have a working example of the model and then we can take it from this example :-) 
Maybe a google colab using this branch of diffusers would be great!

kashif commented:
@patrickvonplaten sure I am working on a colab and will share!

patrickvonplaten commented:
Hey,

I looked into the notebook: https://colab.research.google.com/drive/1ntgPTgR6tQ-PJ14GOmfQM7tVloot8z3S?usp=sharing 

but I don't seem to be able to load it - any ideas what could be the reason here @kashif ? 

Would be amazing if you we could get it to run in a public google colab notebook

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,666 @@
+
+                # Denoising diffusion loop
+                for t in self.progress_bar(self.scheduler.timesteps):
+                    output = self.decode(
```
Nice, so the `decoder` is used as a diffusion decoding model.

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,666 @@
+        num_inference_steps: int = 1000,
+        return_dict: bool = True,
+    ):
+        ns = note_seq.midi_file_to_note_sequence(midi_file)
```
nice! We should be sure to add `note_seq` as an optional dependency

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,666 @@
+        return spec_out
+
+
+class SpectrogramDiffusionPipeline(DiffusionPipeline):
```
This pipeline already looks very nice!

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,666 @@
+        return self.dropout_post(x), encoder_inputs_mask
+
+
+class SpectrogramContEncoder(ModelMixin, ConfigMixin, ModuleUtilsMixin):
```
This model is not a "diffusers" model so happy to move leave it here in the pipelines and depend on `transformers`. Could we maybe add it under `src/diffusers/pipelines/spectrogram_diffusion/encoder.py` ? To make it easier, I'm also happy to have it be a parent class of `PreTrainedModel` just like: https://github.com/huggingface/diffusers/blob/261a448c6aa467810545d87499f6f3bed334754f/src/diffusers/pipelines/alt_diffusion/modeling_roberta_series.py#L59

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,666 @@
+        return outputs  # hidden-states, present_key_value_states, (self-attention position bias), (self-attention weights), (cross-attention position bias), (cross-attention weights)
+
+
+class SpectrogramNotesEncoder(ModelMixin, ConfigMixin, ModuleUtilsMixin):
```
Could we also move this to its own file in the pipelines, just like: https://github.com/huggingface/diffusers/pull/1044/files#r1052776464

patrickvonplaten commented:
@kashif thanks a lot for making all this progress here!

If possible it would be extremely helpful to add a link to a google colab that runs this pipeline in inference and works with all the required dependencies :-) Think this would really help to gauge how long inference takes, what installs are required etc...

kashif commented:
@patrickvonplaten thank you so much! Kindly have a look here: https://colab.research.google.com/drive/1ntgPTgR6tQ-PJ14GOmfQM7tVloot8z3S?usp=sharing

patrickvonplaten commented:
Thanks a lot @kashif, that's super useful. I was able to run the whole notebook and the results sound very nice!

Just two things I noticed:
- 1. The first 5-6 seconds seemed to be just noise for me (think this corresponds exactly to the first generated segment). See generated audio here: https://colab.research.google.com/drive/1W0lBX_PUIcwdp7fCRVWQmTASfRHWo2pi?usp=sharing -> any idea what could be the problem here? 
- 2. The pipeline is really slow. It takes almost 30 mins to generate the audio on a google colab. That's roughly 15x Real-time factor. Do you think there are ways to speed it up? 

Apart from this PR is in a super nice state. If ok for you I could take it over at the end of next week to do some final changes :-)

kashif commented:
thanks, @patrickvonplaten 

1. Regarding this, it is like this because at the start there is no conditioning latents so it generates zeros. Let me double-check the original sampling code to see if they throw this away rather than concating it.

2. yes indeed it is slow, due to the fact we have inference steps of 1000, and then many 5sec segments that get looped over... perhaps another scheduler and fewer inference steps could help?

Thank you for having a look. I will try to investigate issue 1.

kashif commented:
@patrickvonplaten so I believe i fixed issue 1 above with 5 sec of empty noise for the first segment. 

regarding issue 2, i reduced the amount of cpu/gpu transfer now, as well as making repeated ones and zeros tensors, and also fixed the fp16/bf16 inference pipeline... I do not however see much improvement in speed though...

patrickvonplaten commented:
> @patrickvonplaten so I believe i fixed issue 1 above with 5 sec of empty noise for the first segment.
> 
> regarding issue 2, i reduced the amount of cpu/gpu transfer now, as well as making repeated ones and zeros tensors, and also fixed the fp16/bf16 inference pipeline... I do not however see much improvement in speed though...

Awesome! Thanks a lot :-) I will try to take a deeper look into the PR early next week :crossed_fingers:

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -572,7 +572,7 @@ def __call__(
         midi_file,
         generator: Optional[torch.Generator] = None,
-        num_inference_steps: int = 1000,
+        num_inference_steps: int = 100,
```
100 steps is actually just fine in my experiments

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -618,7 +618,7 @@ def __call__(
         ones = torch.ones((1, TARGET_FEATURE_LENGTH), dtype=np.bool, device=self.device)
 
-        for i, encoder_input_tokens in enumerate(input_tokens):
+        for i, encoder_input_tokens in enumerate(input_tokens[:2]):
```
Just for testing for now

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -658,7 +658,7 @@ def __call__(
                     encodings_and_masks=encodings_and_masks,
                     input_tokens=x,
-                    noise_time=t / num_inference_steps,  # rescale to [0, 1)
+                    noise_time=t / self.scheduler.config.num_train_timesteps,  # rescale to [0, 1)
```
This allows us to use less than 1000 steps

patrickvonplaten commented on src/diffusers/models/t5_film_transformer.py:
```diff
@@ -0,0 +1,303 @@
+from .cross_attention import CrossAttention
+
+
+class T5FilmDecoder(ModelMixin, ConfigMixin):
```
The whole T5 Decoder model is now here and uses the optimized `diffusers` cross attention

patrickvonplaten reviewed (commented):
Did some bigger changes, but I think the design now fits quite nicely :-) 
Think I can do some final clean-ups tomorrow and then we should be ready for a final review

patrickvonplaten commented:
@kashif should we try to finish the PR next week? Think the community would be very excited about this :-)

kashif commented:
ok @patrickvonplaten let me refresh it and check again

patrickvonplaten commented on tests/pipelines/spectrogram_diffusion/test_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,68 @@
+    def test_spectrogram(self):
+        device = torch_device
+
+        pipe = SpectrogramDiffusionPipeline.from_pretrained("/home/patrick_huggingface_co/music-spectrogram-diffusion", torch_dtype=torch.float16)
```
```suggestion
        pipe = SpectrogramDiffusionPipeline.from_pretrained("kashif/music-spectrogram-diffusion", torch_dtype=torch.float16)
```

patrickvonplaten commented on tests/pipelines/spectrogram_diffusion/test_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,68 @@
+    def test_spectrogram_fast(self):
+        device = torch_device
+
+        pipe = SpectrogramDiffusionPipeline.from_pretrained("/home/patrick_huggingface_co/music-spectrogram-diffusion")
```
```suggestion
        pipe = SpectrogramDiffusionPipeline.from_pretrained("kashif/music-spectrogram-diffusion", torch_dtype=torch.float16)
```

patrickvonplaten reviewed (commented):
Think this PR is now in a very good state! We should make sure to add some docs for it though :-) 

Also, for this model, adding callbacks will be very important so that when long audio is generated, intermediate values can already be decoded.

patrickvonplaten commented on tests/pipelines/spectrogram_diffusion/test_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,68 @@
+    def test_spectrogram_fast(self):
+        device = torch_device
+
+        pipe = SpectrogramDiffusionPipeline.from_pretrained("kashif/music-spectrogram-diffusion", torch_dtype=torch.float16)
```
```suggestion
        pipe = SpectrogramDiffusionPipeline.from_pretrained("kashif/music-spectrogram-diffusion")
```

patrickvonplaten commented on src/diffusers/pipelines/__init__.py:
```diff
@@ -35,6 +36,14 @@
     from .audio_diffusion import AudioDiffusionPipeline, Mel
 
+try:
+    if not (is_torch_available() and is_note_seq_available()):
```
```suggestion
    if not (is_torch_available() and is_note_seq_available()) and is_onnx_available()):
```

patrickvonplaten commented on docs/source/en/api/pipelines/spectrogram_diffusion.mdx:
```diff
@@ -0,0 +1,52 @@
+pipe = SpectrogramDiffusionPipeline.from_pretrained("kashif/music-spectrogram-diffusion")
+pipe = pipe.to("cuda")
+
+output = pipe("beethoven_hammerklavier_2.mid")
```
Can we also download this midi file in the example?

patrickvonplaten replied:
Can we update the example here before merging?

patrickvonplaten commented on docs/source/en/api/pipelines/spectrogram_diffusion.mdx:
```diff
@@ -0,0 +1,52 @@
+## SpectrogramDiffusionPipeline
+[[autodoc]] SpectrogramDiffusionPipeline
+	- all
+	- __call__
```
We should also add a space here that shows how the model can be used in practice - happy to help :-)

patrickvonplaten commented on src/diffusers/pipelines/__init__.py:
```diff
@@ -35,6 +36,14 @@
     from .audio_diffusion import AudioDiffusionPipeline, Mel
 
+try:
+    if not (is_torch_available() and is_note_seq_available()) and is_onnx_available()):
```
```suggestion
    if not (is_torch_available() and is_note_seq_available())):
```

Sorry upon second reflection let's maybe not make it dependent on ONNX and instead allow to return raw mel-gan optionally. This would also make it much easier to test the model (see comments in pipeline)

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,237 @@
+
+
+class SpectrogramDiffusionPipeline(DiffusionPipeline):
+    def __init__(
```
```suggestion
    _optional_components = ["melgan"]

    def __init__(
```

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,237 @@
+
+from ...models import T5FilmDecoder
+from ...schedulers import DDPMScheduler
+from ...utils import is_note_seq_available, randn_tensor
```
```suggestion
from ...utils import is_note_seq_available, randn_tensor
if is_onnx_available():
```

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,237 @@
+from ...models import T5FilmDecoder
+from ...schedulers import DDPMScheduler
+from ...utils import is_note_seq_available, randn_tensor
+from ..onnx_utils import OnnxRuntimeModel
```
```suggestion
if is_onnx_available():
    from ..onnx_utils import OnnxRuntimeModel
```

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,237 @@
+        midi_file,
+        generator: Optional[torch.Generator] = None,
+        num_inference_steps: int = 100,
+        return_dict: bool = True,
```
```suggestion
        output_type: str = "numpy",
        return_dict: bool = True,
```

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,237 @@
+            full_pred_mel = np.concatenate([full_pred_mel, pred_mel[:1]], axis=1)
+            print("Generated segment", i)
+
+        full_pred_audio = self.melgan(input_features=full_pred_mel.astype(np.float32))
```
```suggestion
        if output_type == "np" and not is_onnx_available():
            raise ValueError("Cannot return output in 'np' format if ONNX is not available. Make sure to have ONNX installed or set 'output_type' to 'mel'.")
        elif output_type == "np" and self.melgan is None:
            raise ValueError("Cannot return output in 'np' format if melgan component is not defined. Make sure to define `self.melgan` or set 'output_type' to 'mel'.")
            
        if output_type == "np":
            output = self.melgan(input_features=full_pred_mel.astype(np.float32))
        else:
            output = full_pred_audio
          
```

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,237 @@
+        full_pred_audio = self.melgan(input_features=full_pred_mel.astype(np.float32))
+
+        if not return_dict:
+            return (full_pred_audio,)
```
```suggestion
            return (output,)
```

patrickvonplaten commented on src/diffusers/pipelines/spectrogram_diffusion/pipeline_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,237 @@
+        if not return_dict:
+            return (full_pred_audio,)
+
+        return AudioPipelineOutput(audios=full_pred_audio)
```
```suggestion
        return AudioPipelineOutput(audios=output)
```

patrickvonplaten reviewed (commented):
Cool! Left some final suggestions - it would be nice if we could optionally allow the user to return the melgan outputs directly. This would also help us a lot with testing. 

Also can we add a couple of audio tests as done here: https://github.com/huggingface/diffusers/blob/main/tests/pipelines/dance_diffusion/test_dance_diffusion.py

patrickvonplaten reviewed (commented):
Let's try to also add some fast tests and make a gradio demo for it :-)

patrickvonplaten reviewed (commented):
Think just the fast tests are missing - then we can look into doing a nice demo!

patrickvonplaten commented on src/diffusers/models/t5_film_transformer.py:
```diff
@@ -17,7 +17,7 @@
 
 from ..configuration_utils import ConfigMixin, register_to_config
-from .cross_attention import CrossAttention
+from ..models.attention_processor import Attention
```
```suggestion
from .attention_processor import Attention
```

patrickvonplaten commented on src/diffusers/models/t5_film_transformer.py:
```diff
@@ -0,0 +1,321 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2023 The HuggingFace Team. All rights reserved.
```

patrickvonplaten commented on tests/pipelines/spectrogram_diffusion/test_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,238 @@
+        device = torch_device
+
+        pipe = SpectrogramDiffusionPipeline.from_pretrained("kashif/music-spectrogram-diffusion", melgan=None)
+        melgan = OnnxRuntimeModel.from_pretrained("kashif/soundstream_mel_decoder")
```
```suggestion
        melgan = pipe.melgan
        pipe.melgan = None
```

patrickvonplaten reviewed (approved):
Cool, I think we're very close to merging this one. 

Only thing left to do is:
- Update notes to adhere to the newest API (e.g. don't pass file name to pipeline, but tokens
- Update tests and docs to use `"google/...." model id instead of `"kashif"` 

Also could you maybe update the example here: https://huggingface.co/google/music-spectrogram-diffusion#example-usage

patrickvonplaten commented on tests/pipelines/spectrogram_diffusion/test_spectrogram_diffusion.py:
```diff
@@ -0,0 +1,239 @@
+
+        audio = output.audios[0]
+        rate = 16_000
+        scipy.io.wavfile.write("/home/patrick_huggingface_co/audios/beet.wav", rate, audio[0])
```
```suggestion
```

kashif commented:
@patrickvonplaten I could not reproduce the one failed mps test...

patrickvonplaten commented:
Think the MPS failure is unrelated - let's merge once the tests now are more or less green :-)
